File: src/main/java/me/austin/queer/gui/Component.java
package me.austin.queer.gui;

import me.austin.queer.util.Util;
import net.minecraft.client.util.math.MatrixStack;

public abstract class Component implements Util {
    protected int x, y;
    protected int width, height;
    protected boolean shown;

    public Component(int x, int y, int width, int height) {
        this.x = x;
        this.y = y;
        this.width = width;
        this.height = height;
    } 

    public abstract void render(MatrixStack stack);
    public abstract void mouseClicked(double mousex, double mousey, int button);
    public abstract String getText();

    public void setPosition(int x, int y) {
        this.x = x;
        this.y = y;
    }

    public void setWidth(int width) {
        this.width = width;
    }

    public void setHeight(int height) {
        this.height = height;
    }

    public boolean isShown() {
        return this.shown;
    }

    public int x() {
        return this.x;
    }

    public int y() {
        return this.y;
    }

    public int width() {
        return this.width;
    }

    public int height() {
        return this.height;
    }
}

File: src/main/kotlin/trans/rights/api/gui/Component.kt
package trans.rights.api.gui

import net.minecraft.client.util.math.MatrixStack

abstract class Component(var xPos: Int, var yPos: Int, var width: Int, var height: Int) {
    fun isOver(mouseX: Int, mouseY: Int) = mouseX > xPos && mouseX < xPos + width && mouseY > yPos && mouseY < yPos + height

    fun setPos(x: Int, y: Int) {
        this.xPos = x
        this.yPos = y
    }

    abstract fun render(stack: MatrixStack)
}
